docs/design_notes/type_promotion.md

```diff
@@ -0,0 +1,722 @@
+  'u8': [0.5, 0], 'u16': [1.5, 0], 'u32': [2.5, 0], 'u64': [3.5, 0],
+  'i8': [0, 1], 'i16': [1, 1], 'i32': [2, 1], 'i64': [3, 1],
+  'f16': [0.5, 2], 'f32': [1.5, 2], 'f64': [2.5, 2],
+  'c64': [2, 3], 'c128': [3, 3],
+}
+fig, ax = plt.subplots(figsize=(6, 5))
+nx.draw(graph, with_labels=True, node_size=1500, node_color='lightgray', pos=pos, ax=ax)
+```
+
++++ {"id": "WVjEMlmm9RS5"}
+
+Notice that this is not (yet) a true lattice: there are many pairs of nodes for which a join does not exist. However, we can think of this as a *partial* lattice, in which some pairs of nodes do not have a defined promotion behavior, and the defined portion of this partial lattice does correctly describe NumPy's array promotion behavior (leaving aside value-dependent semantics mentioned above).
+
+This sets up a nice framework by which we can think about filling-out these undefined promotion rules, by adding connections on this graph. But which connections to add?
+Broadly speaking, we want any additional connections to satisfy a few properties:
+
+1. Connections should maintain the commutativity and associativity of type promotion: that is, the graph should remain a (partial) lattice.
+
+2. Connections should never allow for a loss of data: for example, we should never promote `float` to `int`, as it would discard everything beyond the decimal point.
+
+3. Connections should never allow for a loss of magnitude. For example, the maximum possible `uint32` is twice as large as the maximum possible `int32`, so we should not implicitly promote `uint32` to `int32`.
+
+4. Wherever possible, connections should avoid loss of precision. For example, an `int64` value may have 64 bits of mantissa, so promoting `int64` to `float64` represents a possible loss of precision. However, because the maximum representable float64 is larger than the maximum representable int64, the *magnitude* can still be correctly represented.
+
+5. Wherever possible, connections should avoid unnecessary promotion to larger bit-widths. The purpose of this is to ensure that JAX's implicit promotions remain friendly to accelerator-based workflows, in which users often want to restrict types to 32-bit (or in some cases 16-bit) values.
+
+Each new connection introduces some level of convenience to the user (a new set of types that can interact without explicit casting), but the convenience may becomes too costly if any of the above criteria are violated. Developing a full promotion lattice involves striking a balance between this convenience and this cost.
+
++++ {"id": "GSqwTTS8nYdn"}
+
+## Mixed Promotion: Float and Complex
+
+Let's begin with what is perhaps the easiest case, that of promotion between float and complex values.
+
+Complex numbers are made up of pairs of floating point numbers, and so we have a natural path of promotion between them: cast float to complex while maintaining the width of the real part. In terms of our lattice representation, it would look like this:
```

Comment:
You should probably note that this is in fact not a lattice: the supremum is not defined for all values. So you should probably note that this is a partial lattice and that in the future when you write "lattice" you mean partial lattice.

Reply:
Done.